iceoryx_hoofs/include/iceoryx_hoofs/design_pattern/polymorphic_handler.hpp

```diff
@@ -0,0 +1,126 @@
+/// @note The lifetime of external non-default instances must exceed the lifetime of the PolymorphicHandler.
+/// @note The PolymorphicHandler is guaranteed to provide a valid handler during the whole program lifetime (static).
+/// It is hence not advisable to have other static variables depend on the PolymorphicHandler.
+/// It must be ensured that they are destroyed before the PolymorphicHandler.
+/// @note Hooks must implement
+/// static void onSetAfterFinalize(Interface& /*currentInstance*/, Interface& /*newInstance*/).
+template <typename Interface, typename Default, typename Hooks = detail::DefaultHooks<Interface>>
+class PolymorphicHandler
+{
+    static_assert(std::is_base_of<Interface, Default>::value, "Default must inherit from Interface");
+
+    using Self = PolymorphicHandler<Interface, Default, Hooks>;
+    friend class StaticLifetimeGuard<Self>;
+
+  public:
+    /// @brief set the current singleton instance
+    /// @return the current instance
+    /// @note we cannot be sure to use the current handler unless we call get,
+    /// i.e. a reference obtained from get may reference a previous handler
+    /// (that is still functional but inactive)
+    static Interface& get() noexcept;
+
+    /// @brief set the current singleton instance
+    /// @param handlerGuard a guard to the handler instance to be set
+    /// @return pointer to the previous instance or nullptr if the handler could not be replaced
+    /// @note the handler cannot be replaced if it was finalized
+    /// @note using a guard in the interface prevents the handler to be destroyed while it is used,
+    ///       passing the guard by value is necessary (it has no state anyway)
+    template <typename Handler>
+    static Interface* set(StaticLifetimeGuard<Handler> handlerGuard) noexcept;
+
+    /// @brief reset the current singleton instance to the default instance
+    /// @return pointer to the previous instance
+    static Interface* reset() noexcept;
+
+    /// @brief finalizes the instance, afterwards Hooks::onSetAfterFinalize
+    ///        will be called during the remaining program lifetime
+    ///        when attempting to set or reset the handler
+    static void finalize() noexcept;
+
+    /// @brief returns a lifetime guard whose existence guarantees
+    /// the created PolymorphicHandler singleton instance will exist at least as long as the guard.
+    /// @return opaque lifetime guard object for the (implicit) PolymorphicHandler instance
+    /// @note the PolymorphicHandler will exist once any of the static methods (get, set etc.)
+    ///  are called
+    static StaticLifetimeGuard<Self> guard() noexcept;
+
+  private:
+    // should a defaultHandler be created, this delays its destruction
+    StaticLifetimeGuard<Default> m_defaultGuard;
+    std::atomic_bool m_isFinal{false};
+    std::atomic<Interface*> m_current;
+
+    PolymorphicHandler() noexcept;
+
+    static PolymorphicHandler& instance() noexcept;
```

Comment:
This is kind of inconsistent with the `T& StaticLifetimeGuard::instance()`, which returns a `T&` instead of a `Self&`

Reply:
True. Will look for a reasonable name. Maybe `self`.